### ISSUE TITLE
Implement the delete of webhook for github package 

### ISSUE THREAD

**Issue by csantanapr** [openwhisk/openwhisk#388]
the Git package creates a webhook but is not deleting it. There is a missing correlation between the trigger and the webhook.

**Comment by lzbj** [comment]
@csantanapr , by correlation you mean we can return the id of the created hook and user could delete it with parameter in another command?
I searched and found those two docs:
https://developer.github.com/webhooks/
https://developer.github.com/v3/repos/hooks/

we only support create/delete, it seems ping/edit also doable.

**Comment by csantanapr** [comment]
@lzbj yes meaning that when user deletes the trigger from openwhisk, openwhisk will also go to github and delete the webhook from github. so when creating the trigger the webhook info should be saved.

**Comment by lzbj** [comment]
Ok, got it, thank you.

**Comment by rabbah** [comment]
@ChrBi points out that deleting the feed requires the appropriate webhook. This is related to #471.

**Comment by lzbj** [comment]
what about we list all the webhooks created by the user, and check if the callback url contains the trigger name, we then delete the corresponding webhook. In this way, we don’t have to persist the webhook id somewhere. The bad is we delete all the webhooks with the trigger name.

**Comment by starpit** [comment]
this i think is now blocking ui testing, as i cannot re-test creating a webhook, because github squashes any attempt to create a new webhook with the same parameters.

**Comment by lzbj** [comment]
var request = require('request');

/**
 *   Feed to create a webhook on Github
 *  @param {object} params - information about the trigger
 *  @param {string} repository - repository to create webhook
 *  @param {string} username - github username
 *  @param {string} accessToken - github access token
 *  @param {string} events - list of the events the webhook should fire on
 *  @return {object} whisk async
 */
function main(params) {
    var username = params.username;
    var repository = params.repository;
    var accessToken = params.accessToken;

    var endpoint = 'openwhisk.ng.bluemix.net';
    var lifecycleEvent = params.lifecycleEvent;
    var triggerName = params.triggerName.split("/");

    // URL of the whisk system. The calls of github will go here.
    var whiskCallbackUrl = 'https://' + whisk.getAuthKey() + "@" + endpoint + '/api/v1/namespaces/' + triggerName[1] + '/triggers/' + triggerName[2];

    // The URL to create the webhook on Github
    var registrationEndpoint = 'https://api.github.com/repos/' + username + '/' + repository + '/hooks';

    var authorizationHeader = 'Basic ' + new Buffer(username + ':' +
        accessToken).toString('base64');

    if (lifecycleEvent === 'CREATE') {
        var events = params.events.split(',');

        var body = {
            name: 'web',
            active: true,
            events: events,
            config: {
                url: whiskCallbackUrl,
                content_type: 'json'
            }
        };

        var options = {
            method: 'POST',
            url: registrationEndpoint,
            body: JSON.stringify(body),
            headers: {
                'Content-Type': 'application/json',
                'Authorization': authorizationHeader,
                'User-Agent': 'whisk'
            }
        };

        request(options, function(error, response, body) {
            if (error) {
                whisk.error();
            }
            console.log("Status code: " + response.statusCode);
            whisk.done({ response: body });
        });
    }

    if (lifecycleEvent === 'DELETE') {
        //list all the existing webhooks first.
        var options = {
            method: 'GET',
            url: registrationEndpoint,
            headers: {
                'Content-Type': 'application/json',
                'Authorization': authorizationHeader,
                'User-Agent': 'whisk'
            }
        }

        request(options, function(error, response, body) {
            if (error) {
                whisk.error();
            }

            //body is an array of entities of existing webhooks
            for (i = 0; i < body.length; i++) {
                var pathname = url.parse(body[i].url).pathname;
                //if the url pathname contains the triggerName, then delete the webhook
                if (pathname.search(triggerName) != -1) {
                    var options = {
                        method: 'DELETE',
                        url: body[i].url,
                        headers: {
                            'Content-Type': 'application/json',
                            'Authorization': authorizationHeader,
                            'User-Agent': 'whisk'
                        }
                    }

                    request(options, function(error, response, body) {
                        if (error) {
                            whisk.error(error);
                        } else {
                            console.log("Status code: " + response.statusCode);
                        }
                    })

                }
            }
        });
    }
    return whisk.done();
}

**Comment by lzbj** [comment]
[webhook.js.txt](https://github.com/openwhisk/openwhisk/files/322974/webhook.js.txt)

**Comment by csantanapr** [comment]
@luz Please send a pull request if you have a propose solution.  Attaching code makes it very difficult to review and provide feedback on code.

**Comment by starpit** [comment]
isn't it body[i].config.url? here is an example record in the array that api.github.com returns:
```{
    "type": "Repository",
    "id": 8814032,
    "name": "web",
    "active": true,
    "events": [
      "push"
    ],
    "config": {
      "content_type": "json",
      "url": "https://xxx:yyy@openwhisk.ng.bluemix.net/api/v1/namespaces/uitest3_uitest3/triggers/my git trigger"
    },
    "updated_at": "2016-06-19T23:44:40Z",
    "created_at": "2016-06-19T23:44:40Z",
    "url": "https://api.github.com/repos/testLime/lime/hooks/8814032",
    "test_url": "https://api.github.com/repos/testLime/lime/hooks/8814032/test",
    "ping_url": "https://api.github.com/repos/testLime/lime/hooks/8814032/pings",
    "last_response": {
      "code": 422,
      "status": "misconfigured",
      "message": "Invalid HTTP Response: 403"
    }
  }
```

**Comment by starpit** [comment]
i also notice that `body[i].config.url doesn't URL escape the trigger name. notice that my trigger name has spaces in it...

**Comment by csantanapr** [comment]
@lzbj Are you actively working to fix this bug?

**Comment by lzbj** [comment]
@csantanapr , Yes, I try to propose solution for review, yes, I'll take care next time when I propose code snippet.

@starpit, thanks for your review, I'll propose a pull request try to address it.

**Comment by lzbj** [comment]
please help review this pull request.

https://github.com/openwhisk/openwhisk/pull/742

**Comment by lzbj** [comment]
I'll test with it myself locally deployment, I suffer a slow network.

**Comment by lzbj** [comment]
@starpit @ChrBi , could you please help take a look when you available? Thank you.

**Comment by csantanapr** [comment]
@bjustin-ibm this is one of the github bugs that are important to address soon.

**Comment by bjustin-ibm** [comment]
Now that the GitHub trigger code has moved, this will be resolved by openwhisk/openwhisk-catalog#139.

**Comment by bjustin-ibm** [comment]
Resolved by by openwhisk/openwhisk-catalog#139.
